File: chat-service/src/main/java/com/chatflow/chat/entity/ChatRoom.java
package com.chatflow.chat.entity;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.DynamicInsert;
import org.hibernate.annotations.DynamicUpdate;

import java.time.LocalDateTime;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(of = "id")
@Entity
@DynamicInsert
@DynamicUpdate
@Table(name = "chat_rooms", indexes = {
    @Index(name = "idx_chat_room_created_at", columnList = "created_at DESC"),
    @Index(name = "idx_chat_room_name", columnList = "name")
})
public class ChatRoom {

    @Id
    @Column(length = 50)
    private String id;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(length = 500)
    private String description;

    @Column(length = 7)
    private String color;

    @Column(name = "external_id", unique = true)
    private String externalId;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(name = "room_type", length = 20)
    private RoomType roomType = RoomType.GENERAL;

    @Builder.Default
    @JsonProperty("isPrivate")
    @Column(name = "is_private")
    private boolean isPrivate = false;

    @Column(name = "password")
    private String password;

    @Builder.Default
    @Column(name = "allow_invites")
    private boolean allowInvites = true;

    @Builder.Default
    @Column(name = "participant_count")
    private Integer participantCount = 0;

    @Builder.Default
    @Column(name = "max_participants")
    private Integer maxParticipants = 10;

    @Builder.Default
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    public boolean isFull() {
        int count = participantCount != null ? participantCount : 0;
        int max = maxParticipants != null ? maxParticipants : 10;
        return count >= max;
    }

    public static String nextOverflowName(String baseName, long existingCount) {
        return baseName + "-" + (existingCount + 1);
    }
}

File: frontend/lib/shared/models/chat_room.dart
class ChatRoom {
  final String id;
  final String name;
  final String? description;
  final String? color;
  final String roomType;
  final String? allowedRoles;
  final bool isPrivate;
  final int participantCount;
  final int maxParticipants;
  final String? createdBy;
  final String? createdAt;
  final String? lastMessageAt;
  final String? pinnedMessageId;

  ChatRoom({
    required this.id,
    required this.name,
    this.description,
    this.color,
    this.roomType = 'GENERAL',
    this.allowedRoles,
    this.isPrivate = false,
    required this.participantCount,
    this.maxParticipants = 10,
    this.createdBy,
    this.createdAt,
    this.lastMessageAt,
    this.pinnedMessageId,
  });

  factory ChatRoom.fromJson(Map<String, dynamic> json) {
    return ChatRoom(
      id: json['id']?.toString() ?? json['externalId']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      description: json['description']?.toString(),
      color: json['color']?.toString(),
      roomType: json['roomType']?.toString() ?? 'GENERAL',
      allowedRoles: json['allowedRoles']?.toString(),
      isPrivate: json['isPrivate'] == true || json['private'] == true,
      participantCount: (json['participantCount'] as num?)?.toInt() ?? 0,
      maxParticipants: (json['maxParticipants'] as num?)?.toInt() ?? 10,
      createdBy: json['createdBy']?.toString(),
      createdAt: json['createdAt']?.toString(),
      lastMessageAt: json['lastMessageAt']?.toString(),
      pinnedMessageId: json['pinnedMessageId']?.toString(),
    );
  }

  ChatRoom copyWith({
    String? id,
    String? name,
    String? description,
    String? color,
    String? roomType,
    String? allowedRoles,
    bool? isPrivate,
    int? participantCount,
    int? maxParticipants,
    String? createdBy,
    String? createdAt,
    String? lastMessageAt,
    String? pinnedMessageId,
  }) {
    return ChatRoom(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      color: color ?? this.color,
      roomType: roomType ?? this.roomType,
      allowedRoles: allowedRoles ?? this.allowedRoles,
      isPrivate: isPrivate ?? this.isPrivate,
      participantCount: participantCount ?? this.participantCount,
      maxParticipants: maxParticipants ?? this.maxParticipants,
      createdBy: createdBy ?? this.createdBy,
      createdAt: createdAt ?? this.createdAt,
      lastMessageAt: lastMessageAt ?? this.lastMessageAt,
      pinnedMessageId: pinnedMessageId ?? this.pinnedMessageId,
    );
  }

  bool get isFull => participantCount >= maxParticipants;
  bool get isHandoff => roomType == 'HANDOFF';
}
